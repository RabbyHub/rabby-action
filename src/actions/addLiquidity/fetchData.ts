import PQueue from 'p-queue';
import {
  AddLiquidityRequireData,
  FetchActionRequiredData,
  ParsedActionData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import BigNumber from 'bignumber.js';

export const fetchDataAddLiquidity: FetchActionRequiredData<{
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: any;
  token1: any;
}> = async (options, likeAction) => {
  const { actionData, apiProvider, sender, walletProvider, chainId } = options;

  const action =
    likeAction || (<ParsedActionData<'transaction'>>actionData).addLiquidity;

  if (!action || options.type === 'text') {
    return {};
  }

  const { receiver, exchange_rate, token0, token1 } = action;

  const fair_exchange_rate = new BigNumber(token0.price).div(token1.price);

  const diff = +new BigNumber(exchange_rate)
    .minus(fair_exchange_rate)
    .abs()
    .div(fair_exchange_rate)
    .times(100)
    .toFixed(4);

  const receiverInWallet = await walletProvider.hasAddress(receiver);

  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;
  if (!id) {
    return {};
  }

  const result: AddLiquidityRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: false,
    receiver,
    poolRate: exchange_rate,
    token0,
    token1,
    marketRate: fair_exchange_rate.toFixed(),
    diff,
  };
  const queue = new PQueue();

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;

      result.protocol = contractInfo.protocol;
    }
  });

  queue.add(async () => {
    const hasInteraction = await apiProvider.hasInteraction(
      sender,
      chainId,
      id
    );
    result.hasInteraction = hasInteraction.has_interaction;
  });

  await waitQueueFinished(queue);

  return result;
};
