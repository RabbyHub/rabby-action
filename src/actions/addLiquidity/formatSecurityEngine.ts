import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
