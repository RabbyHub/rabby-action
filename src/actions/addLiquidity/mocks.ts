export const preExecData = {
  abi: {
    func: 'mint',
    params: [
      [
        '0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22',
        '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913',
        10,
        -4970,
        -4960,
        2955759909,
        885199302,
        2940981110,
        880773306,
        '0xE2866F2291B5c5bafD2858B9247f93745108A523',
        1754366820,
        0,
      ],
    ],
  },
  abi_str:
    '{\n    "func": "mint",\n    "params": [\n        [\n            "0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22",\n            "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",\n            10,\n            -4970,\n            -4960,\n            2955759909,\n            885199302,\n            2940981110,\n            880773306,\n            "0xE2866F2291B5c5bafD2858B9247f93745108A523",\n            1754366820,\n            0\n        ]\n    ]\n}',
  balance_change: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 29320,
    gas_limit: 0,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'base',
    chain: 'base',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 3629.55,
    price_24h_change: 0.018385589869839923,
    credit_score: 46503412.60907772,
    is_verified: true,
    is_scam: false,
    is_suspicious: null,
    is_core: true,
    is_wallet: true,
    time_at: null,
    low_credit_score: false,
    amount: 1e-5,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
  },
  type_call: {
    action: 'mint',
    contract: '0x827922686190790b37229fd06084350e74485b72',
    contract_protocol_name: 'Aerodrome V3',
    contract_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/base_aerodrome/f02d753bc321dc8ba480f0424a686482.png',
  },
  trace_id: '663ae6c6c0ac4b1aa5651fadba316de2',
} as any;

export const parseTxData = {
  action: { type: 'cancel_tx' },
  contract_call: null,
  log_id: 441860918,
} as any;

export const txData = {
  chainId: 42161,
  from: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  to: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  value: '0x0',
  data: '0x',
  gas: '0x0',
  maxFeePerGas: '0xb71b00',
  maxPriorityFeePerGas: '0xb71b00',
  nonce: '0x4',
};
