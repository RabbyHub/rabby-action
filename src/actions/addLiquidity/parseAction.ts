import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
