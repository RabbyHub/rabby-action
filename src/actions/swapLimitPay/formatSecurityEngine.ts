import {
  AssetOrderRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineSwapLimitPay: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.swapLimitPay) {
      return {};
    }
    const { receiver } = actionData.swapLimitPay;
    const data = requireData as AssetOrderRequireData;
    return {
      multiSwap: {
        from: data.sender,
        receiver: receiver || '',
        chainId,
        id: data.id,
      },
    };
  };
