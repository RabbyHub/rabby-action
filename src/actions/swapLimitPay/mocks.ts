export const preExecData = {
  abi: {
    func: 'swapOnAugustusRFQTryBatchFill',
    params: [
      [
        37455616131198536,
        100000000000000000000,
        9,
        '2d5abc7230eb420488a5ecad928156d5000000000000000000000000013a9e7c',
        '0x0000000000000000000000000000000000000000',
      ],
      [
        [
          [
            102736910460582217411521356700912238922756658863521543331186950836635519177849,
            1724749194,
            '0x6B175474E89094C44Da98b954EedeAC495271d0F',
            '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
            '0x807cF9A772d5a3f9CeFBc1192e939D62f0D9bD38',
            '0x6A000F20005980200259B80c5102003040001068',
            100010000000000000000,
            37269385465214820,
          ],
          '6772ae1d578d491928b840bbaf409b5676186186e3f3271ca022c9870847541033b8449ed1fbc57dcead16203ef575f36efb3e99b4a2dc2eaed415449ef57f141b',
          37269385465214820,
          '',
          '',
        ],
      ],
      '',
    ],
  },
  abi_str:
    '{\n    "func": "swapOnAugustusRFQTryBatchFill",\n    "params": [\n        [\n            37455616131198536,\n            100000000000000000000,\n            9,\n            "2d5abc7230eb420488a5ecad928156d5000000000000000000000000013a9e7c",\n            "0x0000000000000000000000000000000000000000"\n        ],\n        [\n            [\n                [\n                    102736910460582217411521356700912238922756658863521543331186950836635519177849,\n                    1724749194,\n                    "0x6B175474E89094C44Da98b954EedeAC495271d0F",\n                    "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",\n                    "0x807cF9A772d5a3f9CeFBc1192e939D62f0D9bD38",\n                    "0x6A000F20005980200259B80c5102003040001068",\n                    100010000000000000000,\n                    37269385465214820\n                ],\n                "6772ae1d578d491928b840bbaf409b5676186186e3f3271ca022c9870847541033b8449ed1fbc57dcead16203ef575f36efb3e99b4a2dc2eaed415449ef57f141b",\n                37269385465214820,\n                "",\n                ""\n            ]\n        ],\n        ""\n    ]\n}',
  balance_change: {
    success: true,
    error: null,
    send_token_list: [
      {
        id: 'eth',
        chain: 'eth',
        name: 'ETH',
        symbol: 'ETH',
        display_symbol: null,
        optimized_symbol: 'ETH',
        decimals: 18,
        logo_url:
          'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
        protocol_id: '',
        price: 2685.05,
        price_24h_change: -0.01848216668311636,
        credit_score: 100000000.0,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: 1483200000.0,
        amount: 0.037265658899324886,
        usd_value: 100.06015742763229,
        raw_amount: '37265658899324888',
      },
    ],
    receive_token_list: [
      {
        id: '0x6b175474e89094c44da98b954eedeac495271d0f',
        chain: 'eth',
        name: 'Dai Stablecoin',
        symbol: 'DAI',
        display_symbol: '',
        optimized_symbol: 'DAI',
        decimals: 18,
        logo_url:
          'https://static.debank.com/image/eth_token/logo_url/0x6b175474e89094c44da98b954eedeac495271d0f/549c4205dbb199f1b8b03af783f35e71.png',
        protocol_id: 'makerdao',
        price: 0.9998,
        price_24h_change: 0.0,
        credit_score: 4806813.332173079,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: 1573672677.0,
        amount: 100.0,
        usd_value: 99.98,
        raw_amount: '100000000000000001311',
      },
    ],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: -0.08015742763228673,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 161645,
    gas_limit: 167454,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'eth',
    chain: 'eth',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 2685.05,
    price_24h_change: -0.01848216668311636,
    credit_score: 100000000.0,
    is_verified: true,
    is_scam: false,
    is_suspicious: false,
    is_core: true,
    is_wallet: true,
    time_at: 1483200000.0,
    amount: 0.9797499343379052,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: true,
    error: null,
  },
  type_call: {
    action: 'swapOnAugustusRFQTryBatchFill',
    contract: '0x6a000f20005980200259b80c5102003040001068',
    contract_protocol_name: 'ParaSwap',
    contract_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/paraswap/a9a94be854135149254438560dfa461c.png',
  },
  trace_id: '6ae14127a5a642a3bb07bec9402bc9bd',
} as any;

export const parseTxData = {
  action: {
    type: 'swap_token',
    data: {
      pay_token: {
        id: 'eth',
        chain: 'eth',
        name: 'ETH',
        symbol: 'ETH',
        display_symbol: null,
        optimized_symbol: 'ETH',
        decimals: 18,
        logo_url:
          'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
        protocol_id: '',
        price: 2685.05,
        price_24h_change: -0.01848216668311636,
        credit_score: 100000000.0,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: 1483200000.0,
        amount: 0.037455616131198535,
        raw_amount: '37455616131198536',
      },
      receive_token: {
        id: '0x6b175474e89094c44da98b954eedeac495271d0f',
        chain: 'eth',
        name: 'Dai Stablecoin',
        symbol: 'DAI',
        display_symbol: '',
        optimized_symbol: 'DAI',
        decimals: 18,
        logo_url:
          'https://static.debank.com/image/eth_token/logo_url/0x6b175474e89094c44da98b954eedeac495271d0f/549c4205dbb199f1b8b03af783f35e71.png',
        protocol_id: 'makerdao',
        price: 0.9998,
        price_24h_change: 0.0,
        credit_score: 4806813.332173079,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: 1573672677.0,
        min_amount: 100.0,
        min_raw_amount: '100000000000000000000',
      },
      receiver: '0x341a1fbd51825e5a107db54ccb3166deba145479',
    },
    is_asset_changed: true,
    is_involving_privacy: false,
  },
  contract_call: {
    func: 'swapOnAugustusRFQTryBatchFill',
    contract: {
      id: '0x6a000f20005980200259b80c5102003040001068',
      protocol: {
        name: 'ParaSwap',
        logo_url:
          'https://static.debank.com/image/project/logo_url/paraswap/a9a94be854135149254438560dfa461c.png',
      },
    },
  },
  log_id: 89174533,
} as any;

export const txData = {
  chainId: 1,
  from: '0x341a1fbd51825e5a107db54ccb3166deba145479',
  to: '0x6a000f20005980200259b80c5102003040001068',
  value: '0x8511af3657ee48',
  data: '0xda35bb0d000000000000000000000000000000000000000000000000008511af3657ee480000000000000000000000000000000000000000000000056bc75e2d6310000000000000000000000000000000000000000000000000000000000000000000092d5abc7230eb420488a5ecad928156d5000000000000000000000000013a9e7c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000036000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000020e32307d9d484585f6a59ef1d341a1fbd51825e5a107db54ccb3166deba1454790000000000000000000000000000000000000000000000000000000066cd958a0000000000000000000000006b175474e89094c44da98b954eedeac495271d0f000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000807cf9a772d5a3f9cefbc1192e939d62f0d9bd380000000000000000000000006a000f20005980200259b80c51020030400010680000000000000000000000000000000000000000000000056beae51fd2d100000000000000000000000000000000000000000000000000000084684f01a6b76400000000000000000000000000000000000000000000000000000000000001800000000000000000000000000000000000000000000000000084684f01a6b7640000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000022000000000000000000000000000000000000000000000000000000000000000416772ae1d578d491928b840bbaf409b5676186186e3f3271ca022c9870847541033b8449ed1fbc57dcead16203ef575f36efb3e99b4a2dc2eaed415449ef57f141b00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
  gas: '0x3d52d',
  maxFeePerGas: '0x3b9aca00',
  maxPriorityFeePerGas: '0x3b9aca00',
  nonce: '0x139',
};
