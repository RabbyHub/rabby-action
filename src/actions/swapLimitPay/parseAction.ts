import { SwapLimitPay } from '@rabby-wallet/rabby-api/dist/types';
import { ParseAction } from '../../types';

export const parseActionSwapLimitPay: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'swap_token_limit_pay') {
    return {};
  }

  const {
    pay_token: payToken,
    receive_token: receiveToken,
    receiver,
  } = data.data as SwapLimitPay;

  return {
    swapLimitPay: {
      payToken,
      receiveToken,
      receiver,
    },
  };
};
