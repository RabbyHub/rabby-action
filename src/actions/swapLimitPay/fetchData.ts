import { FetchActionRequiredData } from '../../types';
import { fetchDataSwap } from '../swap/fetchData';

export const fetchDataSwapLimitPay: FetchActionRequiredData = async (
  options
) => {
  if (!options.actionData.swapLimitPay) {
    return {};
  }
  return fetchDataSwap(options, options.actionData.swapLimitPay);
};
