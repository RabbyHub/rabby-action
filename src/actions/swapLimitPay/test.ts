import {
  apiProvider,
  ETH_CHAIN_ID,
  formatProvider,
  SENDER,
  walletProvider,
} from '../../../__mocks__';
import { ParsedTransactionActionData } from '../../types';
import { parseTxData, preExecData, txData } from './mocks';
import { parseActionSwapLimitPay } from './parseAction';
import { fetchDataSwapLimitPay } from './fetchData';
import { formatSecurityEngineSwapLimitPay } from './formatSecurityEngine';
import {
  parseAction,
  fetchActionRequiredData,
  formatSecurityEngineContext,
} from '../..';

/**
 * https://app.odos.xyz/
 * multi from token
 */
test.each([
  [
    parseActionSwapLimitPay,
    fetchDataSwapLimitPay,
    formatSecurityEngineSwapLimitPay,
  ],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('MultiSwap', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'transaction',
    data: parseTxData['action'],
    balanceChange: preExecData.balance_change,
    sender: txData.from,
    preExecVersion: preExecData.pre_exec_version,
    gasUsed: preExecData.gas.gas_used,
    tx: txData,
  }) as ParsedTransactionActionData;
  expect(actionData).toMatchSnapshot('parseActionMultiSwap');

  const requireData = await _fetchData({
    type: 'transaction',
    actionData,
    contractCall: parseTxData.contract_call,
    chainId: ETH_CHAIN_ID,
    sender: SENDER,
    walletProvider,
    tx: txData,
    apiProvider,
  });
  expect(requireData).toMatchSnapshot('fetchDataMultiSwap');

  const ctx = await _format({
    type: 'transaction',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineMultiSwap');
});
