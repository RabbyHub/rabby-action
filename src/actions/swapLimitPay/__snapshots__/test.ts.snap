// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSwap: fetchDataMultiSwap 1`] = `{}`;

exports[`MultiSwap: fetchDataMultiSwap 2`] = `
{
  "bornAt": 1591388241,
  "id": "0x6a000f20005980200259b80c5102003040001068",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": true,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;

exports[`MultiSwap: formatSecurityEngineMultiSwap 1`] = `{}`;

exports[`MultiSwap: formatSecurityEngineMultiSwap 2`] = `
{
  "swap": {
    "chainId": "eth",
    "contractAddress": "0x6a000f20005980200259b80c5102003040001068",
    "from": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "id": "0x6a000f20005980200259b80c5102003040001068",
    "receiveTokenIsFake": false,
    "receiveTokenIsScam": false,
    "receiver": "0x341a1fbd51825e5a107db54ccb3166deba145479",
    "receiverInWallet": true,
    "slippageTolerance": 1.311e-17,
    "usdValuePercentage": -0.005868558288551638,
  },
}
`;

exports[`MultiSwap: parseActionMultiSwap 1`] = `{}`;

exports[`MultiSwap: parseActionMultiSwap 2`] = `
{
  "swap": {
    "balanceChange": {
      "success": true,
      "support": true,
    },
    "minReceive": {
      "amount": 100,
      "chain": "eth",
      "credit_score": 4806813.332173079,
      "decimals": 18,
      "display_symbol": "",
      "id": "0x6b175474e89094c44da98b954eedeac495271d0f",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/eth_token/logo_url/0x6b175474e89094c44da98b954eedeac495271d0f/549c4205dbb199f1b8b03af783f35e71.png",
      "min_amount": 100,
      "min_raw_amount": "100000000000000000000",
      "name": "Dai Stablecoin",
      "optimized_symbol": "DAI",
      "price": 0.9998,
      "price_24h_change": 0,
      "protocol_id": "makerdao",
      "symbol": "DAI",
      "time_at": 1573672677,
    },
    "payToken": {
      "amount": 0.037455616131198535,
      "chain": "eth",
      "credit_score": 100000000,
      "decimals": 18,
      "display_symbol": null,
      "id": "eth",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png",
      "name": "ETH",
      "optimized_symbol": "ETH",
      "price": 2685.05,
      "price_24h_change": -0.01848216668311636,
      "protocol_id": "",
      "raw_amount": "37455616131198536",
      "symbol": "ETH",
      "time_at": 1483200000,
    },
    "receiveToken": {
      "amount": 100,
      "chain": "eth",
      "credit_score": 4806813.332173079,
      "decimals": 18,
      "display_symbol": "",
      "id": "0x6b175474e89094c44da98b954eedeac495271d0f",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/eth_token/logo_url/0x6b175474e89094c44da98b954eedeac495271d0f/549c4205dbb199f1b8b03af783f35e71.png",
      "min_amount": 100,
      "min_raw_amount": "100000000000000000000",
      "name": "Dai Stablecoin",
      "optimized_symbol": "DAI",
      "price": 0.9998,
      "price_24h_change": 0,
      "protocol_id": "makerdao",
      "symbol": "DAI",
      "time_at": 1573672677,
    },
    "receiver": "0x341a1fbd51825e5a107db54ccb3166deba145479",
    "slippageTolerance": 1.311e-17,
    "usdValueDiff": "-0.59020209307462640175",
    "usdValuePercentage": -0.005868558288551638,
  },
}
`;
